tests/csapi/room_ban_test.go

```diff
@@ -0,0 +1,78 @@
+package csapi_tests
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+)
+
+// This is technically a tad different from the sytest, in that it doesnt try to ban a @random_dude:test,
+// but this will actually validate against a present user in the room.
+// sytest: Non-present room members cannot ban others
+func TestNotPresentUserCannotBanOthers(t *testing.T) {
+	deployment := Deploy(t, b.MustValidate(b.Blueprint{
+		Name: "abc",
+		Homeservers: []b.Homeserver{
+			{
+				Name: "hs1",
+				Users: []b.User{
+					{
+						Localpart:   "@alice",
+						DisplayName: "Alice",
+					},
+					{
+						Localpart:   "@bob",
+						DisplayName: "Bob",
+					},
+					{
+						Localpart:   "@charlie",
+						DisplayName: "Charlie",
+					},
+				},
+			},
+		},
+	}))
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	charlie := deployment.Client(t, "hs1", "@charlie:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+	})
+
+	bob.JoinRoom(t, roomID, nil)
+
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type:     "m.room.power_levels",
+		StateKey: b.Ptr(""),
+		Content: map[string]interface{}{
+			"users": map[string]interface{}{
+				charlie.UserID: 100,
```

Comment:
I'm surprised this doesn't throw a 4xx as iirc Alice cannot give herself 0 power?

Reply:
This wasn't intentional, but now that you're pointing it out, it *is* odd that this is somehow allowed.

There's either two things happening here;
- `m.room.power_levels` arent set, and so anyone can send them (according to the spec)
- `m.room.power_levels` is set, and alice (with PL 100) can "demote" herself while also promoting someone else, as the auth check to post a PL is checked before the outcome of that post is considered

I'm curious what the spec says about this though, and this is definitely something that should be clarified.

Reply:
Made an issue; https://github.com/matrix-org/matrix-doc/issues/3545